### PR TITLE
Optimize stack

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented Stack function node to improve stack function. It is two times faster with this test code:
```
import time

import chainer
import cupy
xp = cupy

cupy.cuda.Stream.null.synchronize()
x = chainer.Variable(xp.zeros((32, 256)))
xs = [x] * 100
begin = time.time()
for i in range(1000):
    chainer.functions.stack(xs, axis=1)
cupy.cuda.Stream.null.synchronize()
end = time.time()
print('%f sec' % (end - begin))
```

### CODE REVIEW

File: chainer/functions/array/stack.py [hunk 1]
```diff
@@ -1,4 +1,36 @@
...
+        shape = in_types[0].shape
+        for x_type in in_types[1:]:
+            type_check.expect(
+                x_type.dtype == dtype,
+                x_type.shape == shape,
+            )
+
+    def forward(self, inputs):
+        xp = cuda.get_array_module(*inputs)
+        return xp.stack(inputs, axis=self.axis),
```
Comment (niboshi) [inline chainer/functions/array/stack.py]: numpy.stack is not supported in older version of NumPy.
Please avoid it, by using `numpy.concatenate`, for example.

**Comment by unnonouno** [comment]
i fixed it.

File: chainer/functions/array/stack.py [hunk 2]
```diff
@@ -1,4 +1,41 @@
...
+            )
+
+    def forward(self, inputs):
+        xp = cuda.get_array_module(*inputs)
+        if hasattr(xp, 'stack'):
+            return xp.stack(inputs, axis=self.axis),
+        else:
+            # Old numpy does not have numpy.stack.
+            return xp.concatenate(
+                [xp.expand_dims(x, self.axis) for x in inputs], self.axis)
```
Comment (niboshi) [inline chainer/functions/array/stack.py]: I think a trailing `,` is needed.

File: chainer/functions/array/stack.py [hunk 3]
```diff
@@ -1,4 +1,41 @@
-import chainer
+from chainer import cuda
+from chainer import function_node
+from chainer.functions.array import separate
```
Comment (okuta) [inline chainer/functions/array/stack.py]: Please use `import chainer.functions` to avoid import cycles.

File: chainer/functions/array/stack.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 from chainer import cuda
 from chainer import function_node
-from chainer.functions.array import separate
+from chainer import functions
```
Comment (okuta) [inline chainer/functions/array/stack.py]: Please use `import chainer.functions`.

**Comment by unnonouno** [comment]
The test is passed.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**